src/Strings.h

```diff
@@ -0,0 +1,388 @@
+    IteratorT offset = begin;
+    IteratorT iter = begin;
+    while (iter != end) {
+        if (*iter != isPredicate)
+            offset = iter;
+        ++iter;
+    }
+    return std::string(begin, offset + 1);
+}
+
+template <
+    typename IteratorT,
+    typename PredicateT
+>
+inline std::string Trim(IteratorT begin,
+                        IteratorT end,
+                        PredicateT isPredicate) {
+    std::string stringAfterTrimedLeft(TrimLeft(begin, end, isPredicate));
+    return TrimRight(stringAfterTrimedLeft.begin(), stringAfterTrimedLeft.end(), isPredicate);
+}
+
+template <
+    typename IteratorT,
+    typename IsDelimiterT
+>
+inline void Parse(std::string const &input,
+                  IsDelimiterT const &isDelimiter,
+                  std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    Details::Parse(input.begin(), input.end(), isDelimiter, callback);
+}
+
+template <
+    typename IteratorT,
+    typename RegexT
+>
+inline void ParseRegex(std::string const &input,
+                       RegexT const &regexToken,
+                       std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    Details::ParseRegex(input.begin(), input.end(), regexToken, callback);
+}
+
+template <
+    typename IteratorT,
+    typename IsDelimiterT
+>
+inline void ParseNgramWord(std::string &input,
+                           IsDelimiterT const &isDelimiter,
+                           size_t const &ngramRangeMin,
+                           size_t const &ngramRangeMax,
+                           std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    Details::ParseNgramWord(input, isDelimiter, ngramRangeMin, ngramRangeMax, callback);
+}
+
+template <
+    typename IteratorT,
+    typename IsDelimiterT
+>
+inline void ParseNgramWordCopy(std::string const &input,
+                               IsDelimiterT const &isDelimiter,
+                               size_t const &ngramRangeMin,
+                               size_t const &ngramRangeMax,
+                               std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    //copy string to trim
+    std::string inputCopy(input);
+    Details::ParseNgramWord(inputCopy, isDelimiter, ngramRangeMin, ngramRangeMax, callback);
+}
+
+template <typename IteratorT>
+inline void ParseNgramChar(std::string &input,
+                           size_t const &ngramRangeMin,
+                           size_t const &ngramRangeMax,
+                           std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    Details::ParseNgramChar(input, ngramRangeMin, ngramRangeMax, callback);
+}
+
+template <typename IteratorT>
+inline void ParseNgramCharCopy(std::string const &input,
+                               size_t const &ngramRangeMin,
+                               size_t const &ngramRangeMax,
+                               std::function<void (IteratorT&, IteratorT&)> const &callback) {
```

Comment:
>IteratorT&, IteratorT& [](start = 51, length = 22)

This worries me. Binds to a non-const reference. I need an explanation for this. Usually, iterators are passed by value so the originals are not modified. Especially, the end iterator, why it is a non-const ref?

Reply:
the callback functions are used for string decorator so that modification can apply to original string. I'll add comments then